### PR TITLE
feat: add scan fluidx plate with tecan scanner logic

### PR DESCRIPTION
**Pull request by simbig**
- [x] Added automated tests
- [x] Documented for all relevant versions
- [x] Updated the changelog

**Changes**

- Add class `FluidXPlate` as a container for the scanned rackId and the scanned wells associated with a Coordinate (96-well format) from `mll-lab/microplate`
- Add class `TecanScanner` to parse a raw scan result string into a `FluidXPlate`

### CODE REVIEW

File: .github/workflows/continuous-integration.yml [hunk 1]
```diff
@@ -26,7 +28,94 @@ jobs:
           php-version: 7.4
 
       - name: "Validate composer.json and composer.lock"
-        run: composer validate --strict
+        run: composer validate # --strict TODO reintroduce after https://github.com/thecodingmachine/phpstan-safe-rule/pull/26
```
Comment (spawnia) [inline .github/workflows/continuous-integration.yml]: The PR is merged and released, can this be reverted?

File: .github/FUNDING.yml [hunk 2]
```diff
@@ -1 +1 @@
-github: spawnia
+github: mll-lab
```
Comment (spawnia) [inline .github/FUNDING.yml]: This file should be removed, we currently do not accept sponsors

File: src/FluidXPlate.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+declare(strict_types=1);
```
Comment (spawnia) [inline src/FluidXPlate.php]:
```suggestion
<?php declare(strict_types=1);
```

File: src/FluidXPlate.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    public const FLUIDX_BARCODE_REGEX = /* @lang RegExp */ '/[A-Z]{2}(\d){8}/';
+
+    public string $rackId;
+
+    public function __construct(string $rackId)
+    {
+        $this->rackId = $rackId;
+        if (! (bool) \Safe\preg_match(self::FLUIDX_BARCODE_REGEX, $rackId)) {
+            throw new TecanScanException('Invalid rack id:' . $this->rackId);
+        }
```
Comment (spawnia) [inline src/FluidXPlate.php]:
```suggestion
        if (0 === \Safe\preg_match(self::FLUIDX_BARCODE_REGEX, $rackId)) {
            throw new TecanScanException("Invalid rack ID: {$rackId}");
        }
        $this->rackId = $rackId;
```

- Validate first
- Clarify expected results from preg_match
- Avoid validating `$rackId` and throwing on `$this->rackId`
- proper spacing in message

File: src/FluidXPlate.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        return new CoordinateSystem96Well();
+    }
+
+    /**
+     * @param Coordinate<CoordinateSystem96Well> $coordinate
+     */
+    public function addWell(Coordinate $coordinate, $barcode): void
+    {
+        if (! (bool) \Safe\preg_match(self::FLUIDX_BARCODE_REGEX, $barcode)) {
+            throw new TecanScanException('Invalid tube barcode:' . $barcode);
```
Comment (spawnia) [inline src/FluidXPlate.php]:
```suggestion
        if (0 === \Safe\preg_match(self::FLUIDX_BARCODE_REGEX, $barcode)) {
            throw new TecanScanException("Invalid tube barcode: {$barcode}");
```

File: src/TecanScanner.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+ * The plate scanner on a tecan worktable.
+ */
+class TecanScanner
+{
+    public const NO_READ = 'NO READ';
+
+    public static function parseRawContent(string $rawContent): FluidXPlate
+    {
+        if ('' === $rawContent) {
+            throw new TecanScanException('Der Scan-Inhalt ist leer.');
```
Comment (spawnia) [inline src/TecanScanner.php]: Why mix English and German exceptions?

File: src/TecanScanner.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        $lines = collect(StringUtil::splitLines($rawContent));
```
Comment (spawnia) [inline src/TecanScanner.php]:
```suggestion
        $lines = new Collection(StringUtil::splitLines($rawContent));
```

File: src/TecanScanner.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+        $firstLineWithRackId = $lines->shift();
+
+        if (! is_string($firstLineWithRackId) || ! Str::startsWith($firstLineWithRackId, 'rackid,')) {
+            throw new TecanScanException('Der Scanner lieferte keinen Plattenbarcode zurück.');
+        }
+
+        if (FluidXPlate::coordinateSystem()->positionsCount() !== $lines->count()) {
+            throw new TecanScanException('Der Scan-Inhalt sollte aus ' . FluidXPlate::coordinateSystem()->positionsCount() . ' Barcode-Lines bestehen, besteht aber aus ' . $lines->count());
```
Comment (spawnia) [inline src/TecanScanner.php]: Logic to get the values is duplicated. Semantics would be clarified by extracting to something like `$expectedCount` and `$actualCount`

File: src/TecanScanner.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        $plate = new FluidXPlate(Str::substr($firstLineWithRackId, 7));
+
+        $lines->each(function ($line) use ($plate): void {
```
Comment (spawnia) [inline src/TecanScanner.php]: Simpler and more performant with `foreach`

File: tests/Unit/FluidXPlateTest.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
+<?php
+
+declare(strict_types=1);
```
Comment (spawnia) [inline tests/Unit/FluidXPlateTest.php]:
```suggestion
<?php declare(strict_types=1);
```

Tip: modify your PhpStorm templates

File: tests/Unit/FluidXPlateTest.php [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+namespace Mll\LiquidHandlingRobotics\Tests\Unit;
+
+use Mll\LiquidHandlingRobotics\FluidXPlate;
+use Mll\LiquidHandlingRobotics\TecanScanException;
+use Mll\Microplate\Coordinate;
+use Mll\Microplate\CoordinateSystem96Well;
+use PHPUnit\Framework;
+use TypeError;
+
+class FluidXPlateTest extends Framework\TestCase
```
Comment (spawnia) [inline tests/Unit/FluidXPlateTest.php]:
```suggestion
final class FluidXPlateTest extends TestCase
```

Use import

File: tests/Unit/FluidXPlateTest.php [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        self::assertSame($rackId, $fluidXPlate->rackId);
+        self::assertCount(96, $fluidXPlate->wells());
+        self::assertCount(96, $fluidXPlate->freeWells());
+        self::assertCount(0, $fluidXPlate->filledWells());
+    }
+
+    public function testCanNotAddInvalidBarcode(): void
+    {
+        $barcode = 'testWrongBarcode';
+        $this->expectExceptionObject(new TecanScanException('Invalid tube barcode:' . $barcode));
```
Comment (spawnia) [inline tests/Unit/FluidXPlateTest.php]: Minimize the code that comes after to clarify what exactly triggers the exception

File: tests/Unit/FluidXPlateTest.php [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        $rackId = 'AB12345678';
+        $fluidXPlate = new FluidXPlate($rackId);
+        $fluidXPlate->addWell(Coordinate::fromString('A1', new CoordinateSystem96Well()), $barcode);
+    }
+
+    public function testCanOnlyAddStringAsBarcode(): void
+    {
+        $this->expectException(TypeError::class);
```
Comment (spawnia) [inline tests/Unit/FluidXPlateTest.php]: Not clear where exactly the error is triggered

File: tests/Unit/TecanScannerTest.php [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+        self::assertCount(96, $fluidXPlate->wells());
+        self::assertCount(94, $fluidXPlate->freeWells());
+        self::assertCount(2, $fluidXPlate->filledWells());
```
Comment (spawnia) [inline tests/Unit/TecanScannerTest.php]: Redundant, exact contents are validated later

**Review by spawnia — CHANGES_REQUESTED** [review]

File: src/TecanScanner.php (hunk 9 above)
Comment (simbig) [reply src/TecanScanner.php]: I prefer the functional syntax and do not think that the performance is a limiting factor here.
Anyways I change it, but is now requires and extra line with `assert(is_string($line));`

File: src/FluidXPlate/InvalidRackIdException.php [hunk 15]
```diff
@@ -0,0 +1,11 @@
+<?php declare(strict_types=1);
+
+namespace Mll\LiquidHandlingRobotics\FluidXPlate;
+
+class InvalidRackIdException extends FluidXPlateException
```
Comment (spawnia) [inline src/FluidXPlate/InvalidRackIdException.php]: `final` - missing in more classes

File: tests/Unit/FluidXPlateTest.php [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    public function testCanNotAddInvalidBarcode(): void
+    {
+        $barcode = 'testWrongBarcode';
+        $rackId = 'AB12345678';
+        $fluidXPlate = new FluidXPlate($rackId);
+
+        $this->expectExceptionObject(new InvalidTubeBarcodeException($barcode));
+        $fluidXPlate->addWell(Coordinate::fromString('A1', new CoordinateSystem96Well()), $barcode);
```
Comment (spawnia) [inline tests/Unit/FluidXPlateTest.php]: Does the construction of the coordinate throw or `addWell()`?

File: src/TecanScanner.php (hunk 9 above)
Comment (spawnia) [reply src/TecanScanner.php]: An assertion should not be necessary, given `Collection` is generic.

File: tests/Unit/FluidXPlateTest.php (hunk 16 above)
Comment (simbig) [reply tests/Unit/FluidXPlateTest.php]: yes `FluidXPlate->addWell()` in overriding the `Mircroplate->addWell()` and add a validation if the tube-barcode matches a regex.

File: src/TecanScanner.php (hunk 9 above)
Comment (simbig) [reply src/TecanScanner.php]: Without assertion I am receiving this stan error:

https://github.com/mll-lab/liquid-handling-robotics/runs/7071363370?check_suite_focus=true

Comment (spawnia) [reply src/TecanScanner.php]: Fixed by adding larastan.

File: tests/Unit/FluidXPlateTest.php (hunk 16 above)
Comment (spawnia) [reply tests/Unit/FluidXPlateTest.php]: I localized them further.

**Comment by spawnia** [comment]
Needs https://github.com/mll-lab/microplate/pull/14

**Review by spawnia — APPROVED** [review]

**Pull request merged**